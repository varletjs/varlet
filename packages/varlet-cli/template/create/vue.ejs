<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
