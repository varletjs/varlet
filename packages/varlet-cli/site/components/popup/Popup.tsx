import { defineComponent, watch, Transition, Teleport } from 'vue'
import { props } from './props'
import { useLock } from '../context/lock'
import { useZIndex } from '../context/zIndex'
import { useRouteListener, useTeleport, createNamespace } from '../utils/components'

import '../styles/common.less'
import './popup.less'

const { n, classes } = createNamespace('popup')

export default defineComponent({
  name: 'VarPopup',
  inheritAttrs: false,
  props,
  setup(props, { slots, attrs }) {
    const { zIndex } = useZIndex(() => props.show, 3)
    const { disabled } = useTeleport()

    const hidePopup = () => {
      const { closeOnClickOverlay, onClickOverlay } = props

      onClickOverlay?.()

      if (!closeOnClickOverlay) {
        return
      }

      props['onUpdate:show']?.(false)
    }

    useLock(
      () => props.show,
      () => props.lockScroll
    )

    watch(
      () => props.show,
      (newValue: boolean) => {
        const { onOpen, onClose } = props
        newValue ? onOpen?.() : onClose?.()
      }
    )

    // internal for Dialog
    useRouteListener(() => props.onRouteChange?.())

    const renderOverlay = () => {
      const { overlayClass = '', overlayStyle } = props

      return (
        <div
          class={classes(n('overlay'), overlayClass)}
          style={{
            zIndex: zIndex.value - 1,
            ...overlayStyle,
          }}
          onClick={hidePopup}
        />
      )
    }

    const renderContent = () => {
      return (
        <div
          class={classes(n('content'), 'var-site-elevation--3', n(`--${props.position}`))}
          style={{ zIndex: zIndex.value }}
          {...attrs}
        >
          {slots.default?.()}
        </div>
      )
    }

    const renderPopup = () => {
      const { onOpened, onClosed, show, overlay, transition, position } = props

      return (
        <Transition name="var-site-fade" onAfterEnter={onOpened} onAfterLeave={onClosed}>
          <div class={classes('var-site--box', n())} style={{ zIndex: zIndex.value - 2 }} v-show={show}>
            {overlay && renderOverlay()}
            <Transition name={transition || `var-site-pop-${position}`}>{show && renderContent()}</Transition>
          </div>
        </Transition>
      )
    }

    return () => {
      const { teleport } = props

      if (teleport) {
        return (
          <Teleport to={teleport} disabled={disabled.value}>
            {renderPopup()}
          </Teleport>
        )
      }

      return renderPopup()
    }
  },
})
