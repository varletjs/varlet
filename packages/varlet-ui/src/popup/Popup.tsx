import { defineComponent, watch, Transition, Teleport } from 'vue'
import { props } from './props'
import { useLock } from '../context/lock'
import { useZIndex } from '../context/zIndex'
import { useRouteListener, useTeleport, createNamespace } from '../utils/components'

import '../styles/common.less'
import './popup.less'

const { n, classes } = createNamespace('popup')

export default defineComponent({
  name: 'VarPopup',
  inheritAttrs: false,
  props,
  setup(props, { slots, attrs }) {
    const { zIndex } = useZIndex(() => props.show, 3)
    const { disabled } = useTeleport()

    const hidePopup = () => {
      const { closeOnClickOverlay, onClickOverlay } = props

      onClickOverlay?.()

      if (!closeOnClickOverlay) {
        return
      }

      props['onUpdate:show']?.(false)
    }

    useLock(
      () => props.show,
      () => props.lockScroll
    )

    watch(
      () => props.show,
      (newValue: boolean) => {
        const { onOpen, onClose } = props
        newValue ? onOpen?.() : onClose?.()
      }
    )

    // internal for Dialog
    useRouteListener(() => props.onRouteChange?.())

    const renderOverlay = () => {
      const { overlayClass = '', overlayStyle } = props

      return (
        <div
          class={classes(n('overlay'), overlayClass)}
          style={{
            zIndex: zIndex.value - 1,
            ...overlayStyle,
          }}
          onClick={hidePopup}
        />
      )
    }

    const renderContent = () => {
      return (
        <div
          class={classes(
            n('content'),
            n(`--${props.position}`),
            [props.defaultStyle, n('--content-background-color')],
            [props.defaultStyle, n('$-elevation--3')]
          )}
          style={{ zIndex: zIndex.value }}
          {...attrs}
        >
          {slots.default?.()}
        </div>
      )
    }

    const renderPopup = () => {
      const { onOpened, onClosed, show, overlay, transition, position } = props

      return (
        <Transition name={n('$-fade')} onAfterEnter={onOpened} onAfterLeave={onClosed}>
          <div class={classes(n('$--box'), n())} style={{ zIndex: zIndex.value - 2 }} v-show={show}>
            {overlay && renderOverlay()}
            <Transition name={transition || n(`$-pop-${position}`)}>{show && renderContent()}</Transition>
          </div>
        </Transition>
      )
    }

    return () => {
      const { teleport } = props

      if (teleport) {
        return (
          <Teleport to={teleport} disabled={disabled.value}>
            {renderPopup()}
          </Teleport>
        )
      }

      return renderPopup()
    }
  },
})
