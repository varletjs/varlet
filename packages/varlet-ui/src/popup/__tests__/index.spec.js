import Popup from '..'
import VarPopup from '../Popup'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'

test('test popup plugin', () => {
  const app = createApp({}).use(Popup)
  expect(app.component(Popup.name)).toBeTruthy()
})

const Wrapper = {
  components: {
    [VarPopup.name]: VarPopup,
  },
  props: ['closeOnClickOverlay', 'onOpen', 'onClose', 'onClickOverlay'],
  data: () => ({
    show: false,
  }),
  template: `
    <var-popup v-model:show="show" v-bind="$props">
      default slot content
    </var-popup>
  `,
}

test('test popup show', async () => {
  const wrapper = mount(Wrapper)
  expect(wrapper.find('.var-popup').isVisible()).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setData({ show: true })
  expect(wrapper.find('.var-popup').isVisible()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setData({ show: false })
  expect(wrapper.find('.var-popup').isVisible()).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
})

test('test popup onOpen & onClose', async () => {
  const onOpen = jest.fn()
  const onClose = jest.fn()

  const wrapper = mount(Wrapper, {
    props: {
      onOpen,
      onClose,
    },
  })
  await wrapper.setData({ show: true })
  expect(onOpen).toHaveBeenCalledTimes(1)

  await wrapper.setData({ show: false })
  expect(onClose).toHaveBeenCalledTimes(1)

  wrapper.unmount()
})

test('test popup close on clickOverlay', async () => {
  const onClose = jest.fn()
  const onClickOverlay = jest.fn()

  const wrapper = mount(Wrapper, {
    props: {
      onClose,
      onClickOverlay,
    },
  })

  await wrapper.setData({ show: true })

  await wrapper.find('.var-popup__overlay').trigger('click')
  expect(onClickOverlay).toHaveBeenCalledTimes(1)
  expect(onClose).toHaveBeenCalledTimes(1)

  await wrapper.setData({ show: true })
  await wrapper.setProps({ closeOnClickOverlay: false })
  await wrapper.find('.var-popup__overlay').trigger('click')
  expect(onClickOverlay).toHaveBeenCalledTimes(2)
  expect(onClose).toHaveBeenCalledTimes(1)

  wrapper.unmount()
})

test('test popup z-index', async () => {
  const wrapper = mount(Wrapper)

  await wrapper.setData({ show: true })

  const prevPopupZIndex = window.getComputedStyle(wrapper.find('.var-popup').element).zIndex
  const prevOverlayZIndex = window.getComputedStyle(wrapper.find('.var-popup__overlay').element).zIndex

  await wrapper.setData({ show: false })
  await wrapper.setData({ show: true })
  expect(window.getComputedStyle(wrapper.find('.var-popup').element).zIndex).toBe(String(+prevPopupZIndex + 3))
  expect(window.getComputedStyle(wrapper.find('.var-popup__overlay').element).zIndex).toBe(
    String(+prevOverlayZIndex + 3)
  )

  wrapper.unmount()
})
