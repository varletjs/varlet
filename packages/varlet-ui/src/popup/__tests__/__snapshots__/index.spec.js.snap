// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test popup show 1`] = `
"<transition-stub>
  <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
    <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
    <transition-stub>
      <!---->
    </transition-stub>
  </div>
</transition-stub>"
`;

exports[`test popup show 2`] = `
"<transition-stub>
  <div class=\\"var--box var-popup\\" style=\\"z-index: 2001;\\">
    <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
    <transition-stub>
      <div class=\\"var-popup__content var-popup--center var-popup--content-background-color var-elevation--3\\" style=\\"z-index: 2003;\\"> default slot content </div>
    </transition-stub>
  </div>
</transition-stub>"
`;

exports[`test popup show 3`] = `
"<transition-stub>
  <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
    <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
    <transition-stub>
      <!---->
    </transition-stub>
  </div>
</transition-stub>"
`;
